### PR TITLE
Product QA | Automation Test Creation - LPS-153326 - Allow developers to make a relationship with a system object through API

### PR DESCRIPTION
**Pull request by Hijky**
**Background**
Add a testcase to make a relationship with a system object through API

**Test Plan**
Given a foundation object
Given manyToMany relationship usersFoundations created
Given two foundation entries created
Given two userAccount entries created
When each foundation is related to each userAccount through userAccounts PUT endpoint
Then I can see the entry correctly related in GET request response

Given a foundation object
Given manyToMany relationship usersFoundations created
Given two foundation entries created
Given two userAccount entries created
Given each foundation is related to each userAccount through userAccounts PUT endpoint
When deleting one of the foundation entries
Then I can see only one entry correctly related in GET request response

Given a foundation object
Given manyToMany relationship usersFoundations created
Given two foundation entries created
Given two userAccount entries created
Given each foundation related to each userAccount through userAccounts PUT endpoint
When deleting all of the foundation entries
Then I can see no one entry correctly related in GET request response

Given a foundation object
Given oneToMany relationship usersFoundations created
Given two foundation entries created
Given the userAccount entry created
Given I relate all the foundation entries with the userAccount entry through the userAccount PUT endpoint
When deleting one of the foundation entries
Then I can see only one entry correctly related in GET request response

Given a firstFoundation object
Given oneToMany relationship usersFirstFoundations created
Given a secondFoundation object
Given manyToMany relationship usersSecondFoundations created
Given two foundation entries created
Given the userAccount entry created
When I relate each foundation entry with the userAccount entry through the userAccount PUT endpoint with different relationships
Then I can see the two entries correctly related in GET request response

**JIRA Link:**
https://issues.liferay.com/browse/LPS-153326

### CODE REVIEW

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by Hijky** [comment]
ci:test:headless-frontend-functional:nocompile

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Review by magjed4289 — COMMENTED** [review]
There are errors in the tests, maybe there's a feature flag missing?

**Comment by Hijky** [comment]
Sure, thanks Meg, I forgot to add the feature flag, I will add it.

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+				userAccountId = "${userAccountId1}");
+		}
+	}
+
+	@disable-webdriver = "true"
+	@priority = "4"
+	test CanRelateCustomObjectEntriesToSystemObjectEntries {
+		property portal.acceptance = "true";
+
+		task ("Given a firstFoundation object") {
```
Comment (magjed4289) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase:189]: why don't we use the foundation object here as well? this way we could move all the foundation creation to the setup and we would start this test with creating secondFoundation

**Review by magjed4289 — CHANGES_REQUESTED** [review]

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+			JSONUser.deleteUserByEmailAddress(userEmailAddress = "${userEmailAddress}");
+		}
+	}
+
+	@disable-webdriver = "true"
+	@priority = "4"
+	test CanDeleteCustomObjectEntryRelatedToSystemObjectByManyToMany {
+		property portal.acceptance = "true";
+
+		task ("Given manyToMany relationship usersFoundations created") {
```
Comment (magjed4289) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase:39]: there are still a lot of Given steps that can be moved all together to the setUp.

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+	property portal.release = "true";
+	property portal.upstream = "true";
+	property testray.component.names = "Object";
+	property testray.main.component.name = "Headless";
+
+	setUp {
+		TestCase.setUpPortalInstanceNoSelenium();
+
+		task ("Given a Foundation object") {
+			var objectDefinitionId2 = ObjectDefinitionAPI.createAndPublishObjectDefinition(
```
Comment (yanan) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase:14]: how many object definitions do we need? Two custom object defitions and one system User, right? I'm curious why we use Id2 and Id4 here

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+				name = "usersFoundations",
+				objectDefinitionId1 = "${objectDefinitionId1}",
+				objectDefinitionId2 = "${objectDefinitionId2}",
+				type = "manyToMany");
+		}
+
+		task ("Given oneToMany relationship usersSecondFoundations created") {
+			var relationshipName = ObjectDefinitionAPI.createRelationship(
+				deletionType = "cascade",
+				en_US_label = "UsersSecondFoundations",
```
Comment (yanan) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase:44]: How about UserSecondFoundations for oneToMany?

Comment (yanan) [reply portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase:44]: I'm also thinkg we already have some object definitions created in ObjectDefinitionAPI, is we're not must use the names Foundation and SecondFoundation, how about updating to other names which existing in ObjectDefinitionAPI?

**Comment by Hijky** [comment]
ci:test:relevant

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+	}
+
+	@disable-webdriver = "true"
+	@priority = "4"
+	test CanDeleteCustomObjectEntryRelatedToSystemObjectByManyToMany {
+		property portal.acceptance = "true";
+
+		var responseWithUserAccount = UserAccountAPI.getRelationshipByUserAccountId();
+
+		var userAccountId1 = JSONUtil.getWithJSONPath("${responseWithUserAccount}", "$.items[?(@.givenName=='usergn1')].id");
```
Comment (magjed4289) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessadminuser/MakeARelationshipWithSystemObjectThroughAPI.testcase:89]: I wonder if we can make the userAccountId1 and userAccountId2 values static 🤔 @yanan @Hijky please take a look at `ObjectDefinitionAPI.setUpGlobalobjectDefinitionId();` that we use in the `portal-web/test/functional/com/liferay/portalweb/tests/enduser/headless/AdvancedObjectAPI/SupportManyToManyRelationships.testcase` and check please if we can apply it here as well :)

**Review by magjed4289 — CHANGES_REQUESTED** [review]

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:test:sf

**Comment by magjed4289** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:test:sf

**Comment by magjed4289** [comment]
@Hijky two of the new tests seem to fail, can you take a look?

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-test/src/testFunctional/macros/MakeARelationshipWithSystemObjectThroughAPI.testcase [hunk 6]
```diff
@@ -0,0 +1,251 @@
+@component-name = "portal-headless"
```
Comment (yanan) [inline modules/apps/object/object-rest-test/src/testFunctional/macros/MakeARelationshipWithSystemObjectThroughAPI.testcase:1]: We need to move this file to tests folder.

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:test:headless-functional

**Comment by magjed4289** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:stop

**Comment by magjed4289** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:relevant

**Comment by magjed4289** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:stop

**Comment by magjed4289** [comment]
liferay/ci:test:headless-functional:nocompile

**Comment by magjed4289** [comment]
ci:test:sf

**Comment by magjed4289** [comment]
ci:forward

**Pull request closed without merging**